authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,147 @@
+            self.handle_authentication_error(err, conn)
+
+        try:
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            while True:
+                result_type, result_data = conn.result(ldap_result_id, 0)
+                if result_data:
+                    if result_type == ldap.RES_SEARCH_ENTRY:
+                        result_list.append(result_data)
+                else:
+                    break
+        except ldap.LDAPError as e:
+            err = 'Failed retrieving results according to result_id; {0}'\
+                .format(e)
+            self.handle_authentication_error(err, conn)
+
+        # unbind admin connection.
+        conn.unbind()
+
+        if len(result_list) != 1:
+            err = 'Expecting a single result matching the user with username' \
+                  ' {0}. Found {1}.'.format(who, len(result_list))
+            self.handle_authentication_error(err, conn)
+
+        # extract the user DN from the search result.
+        user_dn = result_list.pop()[0][0]
+
+        # attempt to bind using the matching user DN and provided password.
+        self.try_bind_s(user_dn, password)
+
+    def handle_authentication_error(self, err, conn):
+        conn.unbind()
+        self.lgr.warning(err)
+        raise AuthenticationException(err)
+
+    def try_bind_s(self, who, password):
+            conn = self.synchronous_bind(who, password)
+            conn.unbind()
+
+    def synchronous_bind(self, who, password):
         # initialize connection to the LDAP server
         try:
-            conn = ldap.initialize(self.directory_url)
-        except Exception as e:
-            raise Exception(
-                'Failed to initialize LDAP connection to {0}; {1}'
-                .format(self.directory_url, str(e)))
+            conn = ldap.initialize(self.ldap_url)
+        except ldap.LDAPError as e:
+            err = 'Failed to initialize LDAP connection to {0}; {1}'\
+                .format(self.ldap_url, e)
+            self.lgr.warning(err)
+            raise RuntimeError(err)
 
-        # trying to bind with the given user and password
+        # bind using the provided credentials
         try:
-            conn.bind_s(username, password)
-            return username
-        except Exception as e:
-            raise Exception(
-                'Failed to authenticate user {0}; {1}'
-                .format(username, str(e)))
-        finally:
-            conn.unbind_s()
+            conn.bind_s(who, password)
+        except ldap.INVALID_CREDENTIALS as e:
+            err = 'Invalid credentials specified for user: \'{0}\''\
+                .format(who)
+            self.handle_authentication_error(err, conn)
+        except ldap.LDAPError as e:
+            err = 'Failed to bind user \'{0}\'; {1}' \
+                .format(who, e)
+            self.handle_authentication_error(err, conn)
```

Comment:
the conn.unbind() is actually being done in the handle_authentication_error function